docs/adr/0003_canonicalization_algorithm.md

```diff
@@ -0,0 +1,83 @@
+
+For each transaction, we traverse it's ancestors, stopping when we hit a confirmed transaction or a transaction that conflicts with a confirmed transaction. If a conflict with a confirmed transaction is found, we can mark that transaction and all it's descendants as `not_canonical`. Otherwise, the entire subset will be `canonical`. If we hit a transaction that is anchored in the best chain, we can mark it and all of it's ancestors as `canonical`.
+
+**Pros:**
+* We can efficiently mark large subsets as canonical/not-canonical.
+
+**Cons:**
+* Like option 2, this does not handle transitively-anchored transactions properly.
+
+#### Option 4: Traverse transactions with anchors first.
+
+The algorithm's premise is as follows:
+
+1. If transaction `A` is determined to be canonical, all of `A`'s ancestors must also be canonical.
+2. If transaction `B` is determined to be NOT canonical, all of `B`'s descendants must also be NOT canonical.
+3. If a transaction is anchored in the best chain, it is canonical.
+4. If a transaction conflicts with a canonical transaction, it is NOT canonical.
+5. A transaction with a higher last-seen has precedence.
+6. Last-seen values are transitive. A transaction's real last-seen value is the max between it's last-seen value all of it's descendants.
+
+Like Option 3's algorithm, we maintain two mutually-exclusive `txid` sets: `canoncial` and `not_canonical`.
+
+Imagine a method `mark_canonical(A)` that is based on premise 1 and 2. This method will mark transaction `A` and all of it's ancestors as canonical. For each transaction that is marked canonical, we can iterate all of it's conflicts and mark those as `non_canonical`. If a transaction already exists in `canoncial` or `not_canonical`, we can break early, avoiding duplicate work.
+
+This algorithm iterates transactions in 3 runs.
+
+1. Iterate over all transactions with anchors in descending anchor-height order. For any transaction that has an anchor pointing to the best chain, we call `mark_canonical` on it. We iterate in descending-height order to reduce the number of anchors we need to check against the `ChainOracle` (premise 1). The purpose of this run is to populate `non_canonical` with all transactions that directly conflict with anchored transactions and populate `canonical` with all anchored transactions and ancestors of anchors transactions (transitive anchors).
+2. Iterate over all transactions with last-seen values, in descending last-seen order. We can call `mark_canonical` on all of these that do not already exist in `canonical` or `not_canonical`.
+3. Iterate over remaining transactions that contains anchors (but not in the best chain) and have no last-seen value. We treat these transactions in the same way as we do in run 2.
+
+**Pros:**
+* Transitive anchors are handled correctly.
+* We can efficiently mark large subsets as canonical/non-canonical.
+
+**Cons:** ?
```

Comment:
```suggestion
**Cons:** More API and code changes.
```

Reply:
I don't think the API surface changes are substantial. Definitely more code changes though.